PXT-382: Bug: Similarity queries do not work correctly with ReloadTester

anushas opened the pull request:
When there are multiple embedding indexes on a column, the similarity query must specify the index to use. When such a similarity query with a particular index specified is executed after reloading metadata from persistence via the ReloadTester APIs, it fails to run.

RCA: the index name is passed to the SimilarityExpr as a runtime parameter and is not serialized to persistence as part of the SimilarityExpr. Therefore, on reloading from persistence, the index name information is lost and is set as None. Since there are multiple indexes, the SimilarityExpr fails with an error.

This commit fixes SimilarityExpr to include the index used.

Testing: added coverage for the changes. test repro that failed before, passes with the fix.

aaron-siegel commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/pixeltable/pixeltable/411)
<!-- Reviewable:end -->

mkornacker commented on tests/conftest.py:
```diff
@@ -146,7 +146,22 @@ def img_tbl_exprs(indexed_img_tbl: catalog.Table) -> list[exprs.Expr]:
+    t = indexed_img_tbl
+    return [
         t.img.similarity('red truck'),
+        t.img.similarity('red truck', idx='img_idx0')
```
why not just leave these in img_tbl_exprs?

anushas replied:
because i wanted to also verify they have same id. I have reverted this now that i moved the extra checks to another testcase.
<!-- Reviewable comment -OETQXiF77WrwvSK5iIX:bfvpssk -->
<!-- Sent from Reviewable.io -->

mkornacker commented on tests/conftest.py:
```diff
@@ -146,7 +146,22 @@ def img_tbl_exprs(indexed_img_tbl: catalog.Table) -> list[exprs.Expr]:
+def sim_exprs_diff_idx(multi_idx_img_tbl: catalog.Table) -> list[exprs.Expr]:
+    t = multi_idx_img_tbl
+    return [
+        t.img.similarity('red truck', idx='img_idx1'),
```
why not multi_idx_img_tbl_exprs?

anushas replied:
because it only had similarity exprs. I have changed the name, it fits better. thanks!
<!-- Reviewable comment -OETQn11E0FlAfNZgVlT:b60imrc -->
<!-- Sent from Reviewable.io -->

mkornacker commented on tests/test_exprs.py:
```diff
@@ -865,41 +865,44 @@ def test_similarity2(self, img_tbl: catalog.Table) -> None:
             assert e.id is not None
             d[e.id] = e
-        assert len(d) == len(test_tbl_exprs) + len(img_tbl_exprs)
+        assert len(d) == nexprs_before + 1
```
you're testing something here that wasn't part of the original test (its intention), which is that the two similarity exprs are identical. that should go in a different test case in test_index.

anushas replied:
that's a good idea. since it is focused on the expression, i moved it to test_similarity inside test_expr.py. hope you agree.
i initially thought it was ok to extend this test case as it was already verifying each id was unique. but its better separate, thanks!
<!-- Reviewable comment -OETQx0c07zCf0rLAV1Q:byeszxc -->
<!-- Sent from Reviewable.io -->

mkornacker commented on tests/test_exprs.py:
```diff
@@ -865,41 +865,44 @@ def test_similarity2(self, img_tbl: catalog.Table) -> None:
-            e_deserialized = Expr.deserialize(e_serialized)
-            assert e.equals(e_deserialized)
+        serialized_vals = set(e.serialize() for e in sim_exprs_same_idx)
+        assert len(serialized_vals) == 1
```
Also something that belongs in a different test case.

anushas replied:
Done.
<!-- Reviewable comment -OETcGBe-8-o8JJvTfdy:b-896fix -->
<!-- Sent from Reviewable.io -->

anushas reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/411)* status: 0 of 5 files reviewed, 4 unresolved discussions (waiting on @mkornacker)



<!-- Sent from Reviewable.io -->

anushas commented:
@mkornacker addressed the comments, and also rebased to get the drop index bug fix and enable the commented test scenario for this one.  PTAL. Sorry had to force push to refresh PR.

mkornacker commented on tests/test_exprs.py:
```diff
@@ -864,42 +864,57 @@ def test_similarity2(self, img_tbl: catalog.Table) -> None:
+        assert sim1.id == sim2.id
+        assert sim1.serialize() == sim2.serialize()
+        # index name should be present in the serialized expression
+        assert 'img_idx0' in sim1.serialize()
```
That's internal to the implementation and shouldn't be part of the test.

anushas replied:
Okay, removed.
<!-- Reviewable comment -OEr2XfLB-cnGivWdzAm:b-bl9t1h -->
<!-- Sent from Reviewable.io -->

mkornacker commented on tests/test_exprs.py:
```diff
@@ -864,42 +864,57 @@ def test_similarity2(self, img_tbl: catalog.Table) -> None:
+        assert 'img_idx0' in sim2.serialize()
+
+        t2 = multi_idx_img_tbl
+        # for a table with multiple embedding index, the index
```
grammar

anushas replied:
Corrected. Also corrected a typo.
<!-- Reviewable comment -OEr3qNV3ps7-1yfUkCz:b-lykzfz -->
<!-- Sent from Reviewable.io -->

mkornacker commented on tests/test_index.py:
```diff
@@ -27,28 +27,63 @@ def bad_embed(x: str) -> str:
+        # After the query is serialized, dropping and recreating the index should work
+        # on reload, because the index is available again even if it is not the exact
+        # same one.
+        # recreate the index
```
stray line?

anushas replied:
Removed.
<!-- Reviewable comment -OEr4MAV7yVc8Dwn2avF:bss4a2b -->
<!-- Sent from Reviewable.io -->

anushas reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/pixeltable/pixeltable/411)* status: 0 of 5 files reviewed, 7 unresolved discussions (waiting on @aaron-siegel and @mkornacker)



<!-- Sent from Reviewable.io -->
